# ch08. 예외처리(Exception Handling)
## 프로그램 오류
* 발생 상황에 따라 분류
  * 컴파일 에러 : 컴파일 할 때 발생하는 오류
  * 런타임 에러 : 실행 중간 발생하는 오류
  * 논리적 에러 : 실행 문제 없이 되는데 의도랑 다르게 작동하는 거
* 에러(error)는 프로그램 로직으로 수습 안되는 오류 / 예외(exception)은 코드 로직으로 수습되는 오류
* 쉽게 정의하면 에러는 프로그래머 실수. 예외는 사용자 실수

## 예외 클래스의 계층구조
* Throwable 하위에 Exception / Error 클래스 정의
  * Excetion과 Error 클래스 하위에 다양한 예외, 에러들 정의함~

## 예외처리하기 - try-catch문
* 예외처리는 프로그램 실행 도중 발생할 수 있는 예외 상황을 사전에 처리하는 것이고, 이를 통해 프로그램이 매끄럽게 동작할 수 있도록(정상적인 상태 유지할 수 있도록) 하는거~
* 예외를 처리하기 위해 try-catch 문을 사용한다!

## try-catch문에서의 흐름
* try문에서 예외 발생 시, 발생한 예외에 맞는 catch문을 찾아가 수행(없으면 예외처리 안됨)

## 예외의 발생과 catch블럭
* 검사 결과가 true인 catch블럭은 모두 수행함
* `|` 를 이용하여 멀티캐치블럭 만들 수 있음

## 메서드에 예외 선언하기
* 메서드 선언부에 throw를 사용해서 메서드 내에서 발생할 수 있는 예외를 적어주면 메서드에 예외를 선언할 수 잇음
* 역할 : 해당 메서드 사용할 때 어떤 예외들 처리되어야 하는지 알 수 있음
* 예외가 발생한 메서드에서 예외 처리를 하지 않고 해당 메서드를 호출한 메서드에 넘겨 줄 수는 있음. 그래도 어딘가에서는 try-catch로 예외처리 해주긴 해야 함...

## finally블럭
* 예외 발생 여부와 상관 없이 실행되어야 할 코드
* try -> catch -> finally 순으로 실행됨

## 자동 자원 반환 - try-with-resources문
* try-catch문의 변형
* 파일 관련에서, finally에서 file close 로직을 담을 때 애로사항을 해결할 수 있는 코드
* try() 괄호 안에 객체를 생성하는 문장을 넣으면 try를 벗어나는 순간 자동으로 close()가 호출됨

## 사용자정의 예외 만들기
## 예외 되던지기(exception re-throwing)
* 한 메서드에서 발생 할 수 있는 예외가 여럿이면 일부는 try-catch문으로 처리하고 일부는 선언부에 지정해서 호출한 메서드에서 처리하게 함으로서 나눠서 처리
* 하나의 예외에 대해서도 예외 발생 메서드와 호출 메서드 양쪽에서 처리 하도록 가능 -> 이걸 예외 내던지기라고 함
* catch문에서 필요 작업 한 다음에 throw로 예외 다시 발생시키는거~

## 연결된 예외(chained exception)
* 한 예외가 다른 예외를 발생시키는 경우
* 예외가 원인예외를 포함하도록 함...

## Q)
* 어떤 경우에 try-catch를 사용하지 않고 메서드에 예외를 선언하도록 할까요? 생각나는 예시를 들어주세요
